### PR TITLE
Add new field for DWOC for DevWorkspace pod's runtimeClass

### PR DESCRIPTION
**Pull request by dkwon17**
### What does this PR do?
Introduces a new field for the DWOC for the DevWorkspace pod's runtimeClass

### What issues does this PR fix or reference?
https://github.com/devfile/devworkspace-operator/issues/1286

### Is it tested? How?

To test this PR,

1. Create the plain-devworkspace DevWorkspace on the cluster
```
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: plain-devworkspace
spec:
  started: true
  routingClass: 'basic'
  template:
    components:
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryRequest: 256Mi
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
```

2. Verify that the workspace pod starts and runs properly.

3. Create this DWOC:
```
apiVersion: controller.devfile.io/v1alpha1
config:
  workspace:
    runtimeClassName: test
kind: DevWorkspaceOperatorConfig
metadata:
  name: devworkspace-operator-config
  namespace: openshift-operators
```

4. Restart the plain-devworkspace DevWorkspace.
The DevWorkspace startup should fail with:
```
'Error creating DevWorkspace deployment: Detected unrecoverable deployment condition: FailedCreate pods "workspace9b9d2d153f3747de-55bb88dd5b-" is forbidden: pod rejected: RuntimeClass "test" not found'
```

5. To verify that the runtimeClassName attribute takes precedence over the DWOC-defined runtimeClassName, create the following DevWorkspace:
```
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: plain-devworkspace-2
spec:
  started: true
  routingClass: 'basic'
  template:
    attributes:
      controller.devfile.io/runtime-class: dw-class
    components:
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryRequest: 256Mi
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
```

6. The DevWorkspace startup should fail with:
```
'Error creating DevWorkspace deployment: Detected unrecoverable deployment condition: FailedCreate pods "workspaceb61b62899c1549aa-55bcc6dbfc-" is forbidden: pod rejected: RuntimeClass "dw-class" not found'
```

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

File: controllers/workspace/util_test.go
Comment (AObuchow) [inline controllers/workspace/util_test.go:272]: This function takes the `name` but doesn't actually use it. Also, maybe we should capitalize the abbreviation `Dwoc` -> `DWOC`?

File: controllers/workspace/util_test.go
Comment (AObuchow) [inline controllers/workspace/util_test.go:276]: You can use the `testNamespace` constant instead since it's already defined

File: controllers/workspace/testdata/test-devworkspace-config.yaml
Comment (AObuchow) [inline controllers/workspace/testdata/test-devworkspace-config.yaml:12]: Are you specifically trying to make sure this works with an external DWOC instead of the global DWOC? If so, it might be worthwhile to use a different name for the DWOC, since this is the name of the global DWOC.

File: controllers/workspace/devworkspace_controller_test.go
Comment (AObuchow) [inline controllers/workspace/devworkspace_controller_test.go:669]: I suggest using an `Eventually()` here so that we can set a timeout for the updateDevWorkspace function, similar to what we are [doing](https://github.com/devfile/devworkspace-operator/blob/8fc183f55f20e6854f220a2c4bb844190366f67a/controllers/workspace/devworkspace_controller_test.go#L634) elsewhere.

Additionally, since `updateDevWorkspace()` is not called anywhere else, you could just write it out inline rather than have a function for it. Though, having a function that you call after modifying the devworkspace seems to make sense :)

File: controllers/workspace/util_test.go
Comment (AObuchow) [inline controllers/workspace/util_test.go:61]: If you keep this function and update it to use a timeout, you might also want to consider taking in an error message string that will be given if the update to the devworkspace fails.

File: controllers/workspace/testdata/test-devworkspace-config.yaml
Comment (AObuchow) [inline controllers/workspace/testdata/test-devworkspace-config.yaml:1]: The name of this file is a bit surprising -- I would expect it to actually contain a DWOC, not a devworkspace.
If the intention of this DevWorkspace is to specifically use a DWOC, maybe we should call it something like `test-devworkspace-using-config`?

Though, shouldn't devworkspaces use the global config anyhow?  We configure the global DWOC for testing [here](https://github.com/devfile/devworkspace-operator/blob/8fc183f55f20e6854f220a2c4bb844190366f67a/controllers/workspace/devworkspace_controller_test.go#L745-L750).

If you're specifically trying to test using the external DWOC, it might be more appropriate to name this file something like `test-devworkspace-external-config`.

**Review by AObuchow — COMMENTED** [review]
Looks great to me so far, glad to see you're adding DevWorkspace Controller tests :grin: Left some minor comments.

File: controllers/workspace/util_test.go
Comment (dkwon17) [reply controllers/workspace/util_test.go]: I ended up removing this function since there is the `config.SetGlobalConfigForTesting` function

File: controllers/workspace/testdata/test-devworkspace-config.yaml
Comment (dkwon17) [reply controllers/workspace/testdata/test-devworkspace-config.yaml]: I wanted to use the global DWOC, I removed this .yaml in my latest commit

File: controllers/workspace/devworkspace_controller_test.go
Comment (dkwon17) [reply controllers/workspace/devworkspace_controller_test.go]: Updated, and removed `updateDevWorkspace()`

File: controllers/workspace/testdata/test-devworkspace-config.yaml
Comment (dkwon17) [reply controllers/workspace/testdata/test-devworkspace-config.yaml:1]: My latest commit removes this file

**Comment by dkwon17** [comment]
Thank you @AObuchow , I've updated this PR

File: controllers/workspace/devworkspace_controller_test.go
Comment (AObuchow) [inline controllers/workspace/devworkspace_controller_test.go:668]: Since we're adding an attribute, should we make the error message say: "DevWorkspace should get `runtime-class: test-runtime-class-attribute` attribute" (instead of annotation)?

File: controllers/workspace/devworkspace_controller_test.go
Comment (AObuchow) [inline controllers/workspace/devworkspace_controller_test.go:663]: Here as well, we should probably use "attribute" instead of "annotation"

**Review by AObuchow — COMMENTED** [review]
I still have to do a bit of manual testing, but this looks great to me :) I left some very minor comments.

**Comment by AObuchow** [comment]
@musienko-maxim Any idea what's going wrong with the CI prow tests here?

**Review by AObuchow — APPROVED** [review]
Tested on OpenShift 4.16 and works as expected, great work @dkwon17 :grin: :pray: 
My last suggestion is to squash your feedback commits, and maybe rename the `"Add tests"` commit to something a bit more specific like `"Adds controller environment tests for pod runtimeClass"`

**Comment by dkwon17** [comment]
@AObuchow thanks, I've squashed the commits

**Comment by AObuchow** [comment]
@dkwon17 Awesome, this looks good to merge :) Once all the CI checks pass, feel free to merge :)

**Pull request merged**
